examples/mem-events.rs

```diff
@@ -0,0 +1,138 @@
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+use std::time::Instant;
+
+use clap::{App, Arg, ArgMatches};
+use colored::*;
+use env_logger;
+use kvmi::KVMiPageAccess;
+
+use microvmi::api::{Access, EventReplyType, EventType, InterceptType, Introspectable};
+
+const PAGE_SIZE: usize = 4096;
+
+fn parse_args() -> ArgMatches<'static> {
+    App::new(file!())
+        .version("0.1")
+        .about("Watches memory VMI events")
+        .arg(Arg::with_name("vm_name").index(1).required(true))
+        .get_matches()
+}
+
+fn toggle_pf_intercept(drv: &mut Box<dyn Introspectable>, enabled: bool) {
+    drv.pause().expect("Failed to pause VM");
+
+    let intercept = InterceptType::Pagefault;
+    let status_str = if enabled { "Enabling" } else { "Disabling" };
+    println!("{} memory events", status_str);
+    for vcpu in 0..drv.get_vcpu_count().unwrap() {
+        drv.toggle_intercept(vcpu, intercept, enabled)
+            .expect(&format!("Failed to enable page faults"));
+    }
+
+    drv.resume().expect("Failed to resume VM");
+}
+
+fn handle_pf_event(drv: &mut Box<dyn Introspectable>, gpa: u64, access: u8) {
+    let read_permission = KVMiPageAccess::PageAccessR as u8;
```

Comment:
You are breaking the abstraction here. The example should be hypervisor agnostic.